File: problems/Baekjoon_Online_Judge/Step_by_step/21_Queue_and_Deque/10866_덱.js
/** @param {Input} input */
function solution(input) {
  const N = input.readInt();
  const dequeue = new Deque(N);

  for (let i = 0; i < N; i++) {
    const inputs = input.readStr().split(" ");
    const { command, X } = splitInput(inputs);
    dequeue[command](X);
  }
}

/** @param {string} inputs */
function splitInput(inputs) {
  const command = inputs[0];

  let X = null;
  if (inputs.length == 2) {
    X = parseInt(inputs[1]);
  }

  return { command, X };
}

class Deque {
  dequeue = Array();
  max = 0;
  head = -1;
  tail = -1;
  count = 0;

  constructor(N) {
    this.dequeue = Array(N);
    this.max = N;
  }

  push_front(X) {
    // 앞으로 이동
    this.head -= 1;
    // 넣기 전 인덱스 유효해야 함
    this.head = this.#adjust_index(this.head);

    // 넣기
    this.dequeue[this.head] = X;
    this.count += 1;

    // 다른 쪽 끝도 위치 보장
    this.#sync_index(this.head);
  }

  push_back(X) {
    this.tail += 1;
    this.tail = this.#adjust_index(this.tail);

    this.dequeue[this.tail] = X;
    this.count += 1;

    this.#sync_index(this.tail);
  }

  pop_front() {
    if (this.count > 0) {
      // 제거
      // push 단계에서 유효한 위치를 보장하므로,
      // pop 때는 바로 제거 가능
      const X = this.dequeue[this.head];

      // 뒤로 이동
      this.head += 1;
      this.count -= 1;

      // 이동 후 위치 조정
      this.head = this.#adjust_index(this.head);

      // 다른 쪽도 위치 조정
      this.#sync_index(this.head);

      console.log(X);
    } else {
      console.log(-1);
    }
  }

  pop_back() {
    if (this.count > 0) {
      const X = this.dequeue[this.tail];

      this.tail -= 1;
      this.count -= 1;

      this.tail = this.#adjust_index(this.tail);
      this.#sync_index(this.tail);

      console.log(X);
    } else {
      console.log(-1);
    }
  }

  #adjust_index(index) {
    if (index < 0) {
      index = this.max - 1;
    } else if (index > this.max - 1) {
      index = 0;
    }
    return index;
  }

  #sync_index(index) {
    if (this.count == 1) {
      this.head = index;
      this.tail = index;
    }
  }

  size() {
    console.log(this.count);
  }

  empty() {
    if (this.count == 0) {
      console.log(1);
    } else {
      console.log(0);
    }
  }

  front() {
    if (this.count > 0) {
      console.log(this.dequeue[this.head]);
    } else {
      console.log(-1);
    }
  }

  back() {
    if (this.count > 0) {
      console.log(this.dequeue[this.tail]);
    } else {
      console.log(-1);
    }
  }
}

class Input {
  // 기본값은 파일을 통해 입력받음. OJ 제출 시에는 false로 바꿔야 함!
  #inputByFile = true;
  // 입력 텍스트 파일 이름은 {현재 파일 이름}_input.txt 형식임
  #INPUT_FILE_SUFFIX = ".txt";
  #inputFileName = "";
  // process.stdin은 터미널 통해 입력하는 걸 의미
  #inputStream = process.stdin;
  #inputs = [];

  constructor() {
    if (this.#inputByFile) {
      this.#makeInputFileName();
      this.#setInputStreamAsFile();
    }
    this.#run();
  }

  #makeInputFileName() {
    const path = require("path");
    // __filename은 special attribute 중 하나이며 전체 경로 포함.
    // basename 결과는 파일명만 (+확장자 포함).
    const fileName = path.basename(__filename);
    const fileNameRoot = fileName.split(".")[0];
    this.#inputFileName = `${fileNameRoot}${this.#INPUT_FILE_SUFFIX}`;
  }

  #setInputStreamAsFile() {
    const fs = require("fs");
    const fileStream = fs.createReadStream(this.#inputFileName);
    this.#inputStream = fileStream;
  }

  #run() {
    const readline = require("readline");
    const reader = readline.createInterface({
      input: this.#inputStream,
      output: process.stdout,
    });

    reader
      .on("line", (line) => {
        this.#push(line);
      })
      .on("close", () => {
        // 입력과 출력 사이를 띄우기 위해 한 줄 추가
        console.log("\n");

        solution(this);
        // process.exit();
      });
  }

  #push(element) {
    this.#inputs.push(element);
  }

  readInt() {
    return parseInt(this.#inputs.shift());
  }

  /** @returns {Array<number>} */
  readIntArr() {
    return this.#inputs
      .shift()
      .split(" ")
      .map((number) => parseInt(number));
  }

  /** @returns {string} */
  readStr() {
    return this.#inputs.shift().trim();
  }
}

new Input();

File: problems/Baekjoon_Online_Judge/Step_by_step/21_Queue_and_Deque/10866_덱.py
import sys
from pathlib import Path

input_txt = Path(__file__).with_suffix(".txt")
sys.stdin = open(input_txt)
input = sys.stdin.readline


def main():
    N = int(input())
    deque = Deque(N)

    for _ in range(N):
        inputs = input().split()
        func = deque.__getattribute__(inputs[0])
        arg = inputs[1] if len(inputs) == 2 else None

        result = func(arg)
        # ! result가 None이 아닌지 명시해서 판단하기
        # result가 0일 때 출력 안 될 수 있음.
        print(result) if result is not None else None


class Deque:
    def __init__(self, N: int) -> None:
        self.deque = [0] * N
        self.max_size = N
        self.count = 0
        self.head = -1
        self.tail = -1

    def push_front(self, elem: int) -> None:
        # 앞으로 이동
        self.head -= 1

        # 이동 후 인덱스 유효한지 체크
        if self.head < 0:
            self.head = self.max_size - 1

        # 넣기
        self.deque[self.head] = elem
        self.count += 1

        # 1개만 있을 때는 처음-끝 모두 같음
        if self.count == 1:
            self.tail = self.head

    def push_back(self, elem: int) -> None:
        self.tail += 1

        if self.tail > self.max_size - 1:
            self.tail = 0

        self.deque[self.tail] = elem
        self.count += 1

        if self.count == 1:
            self.head = self.tail

    def pop_front(self, _=None) -> int:
        # 비어있으면 pop 불가
        if self.empty():
            return -1

        # 빼기
        elem = self.deque[self.head]
        self.count -= 1

        # 뒤로 이동
        self.head += 1

        # 이동 후 인덱스 유효한지 체크
        if self.head > self.max_size - 1:
            self.head = 0

        # 1개인 경우 처음-끝 모두 같아야 함
        if self.count == 1:
            self.tail = self.head

        return elem

    def pop_back(self, _=None) -> int:
        if self.empty():
            return -1

        elem = self.deque[self.tail]
        self.count -= 1

        self.tail -= 1

        if self.tail < 0:
            self.tail = self.max_size - 1

        if self.count == 1:
            self.head = self.tail

        return elem

    def size(self, _=None) -> int:
        return self.count

    def empty(self, _=None) -> int:
        return 1 if self.count == 0 else 0

    def front(self, _=None) -> int:
        if self.empty():
            return -1

        return self.deque[self.head]

    def back(self, _=None) -> int:
        if self.empty():
            return -1

        return self.deque[self.tail]


main()
